### ISSUE TITLE
Missing support of SMARTS component-level grouping

### ISSUE THREAD

**Issue by eloyfelix** [rdkit/rdkit#1261]
SMARTS with component-level grouping fail to compile:

`smarts = '(C(=O)O[C,H1]).(C(=O)O[C,H1]).(C(=O)O[C,H1])'`
`patt = Chem.MolFromSmarts(smarts)`
`SMARTS Parse Error: syntax error for input: '(C(=O)O[C,H1]).(C(=O)O[C,H1]).(C(=O)O[C,H1])'`

 http://daylight.com/dayhtml_tutorials/languages/smarts/index.html (6. Component-Level Grouping )

**Comment by greglandrum** [comment]
The inability to parse these is a known gap in the SMARTS parser. It's easy to fix, but the substructure matching algorithm would not do what you expect it to do: it would not obey the restriction that the matches must occur in different fragments.

It is possible to modify the matcher to also handle this, but I have never done so because I have yet to see a real-world use case where this is important functionality and the problem cannot be solved by just having separate molecules. I'd be happy to learn about such a case if it exists.

**Comment by sroughley** [comment]
I also have problems with e.g.
```
smarts='(F.Cl)'
patt=Chem.MolFromSmarts(smarts)
```
whereby I want to indicate the two atoms _are_ in the same component, which I think is definitely a useful SMARTS query language feature?
Steve

**Comment by greglandrum** [comment]
It is indeed a feature of SMARTS, but I'm not convinced about how useful it is.
Do you have a real use case for it that cannot be easily solved by splitting the molecule into separate molecules?

**Comment by sroughley** [comment]
In my case, I'm trying to build a SMARTS from user input of 3 separate atom SMARTS, with an option that one of the atoms is not directly bonded to the other 2, but in that case it still absolutely must be in the same component, so I would combine them something like `({SMARTS_ATOM_0}-!@{SMARTS_ATOM_1}.{SMARTS_ATOM_2})`
I'm sure I could do things to ensure that ATOM 2 is in the same molecule, but the SMARTS expression seems easiest?

**Comment by sroughley** [comment]
That's a fair point @greglandrum - I can't think of a corner case in my use case at the moment where that is not the case.  Is there a simpler / more direct test for "multi-component-ness" than what I normally use (in the Java api, `if(mol.MolToSmiles().contains(".")){...}`)?

**Comment by greglandrum** [comment]
Yep. Either form of `getMolFrags()` returns the number of fragments present.
